### PR TITLE
(SERVER-2565) Write files atomically in CA

### PR DESCRIPTION
**Pull request by pcarlisle**
When writing files always use the atomic-write function from kitchensink which
ensures the file is either fully written or not at all by writing out a temp
file and moving it into place.

### CODE REVIEW

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 1]
```diff
@@ -423,6 +432,51 @@
...
+
+(schema/defn write-private-key
+  "Encode a key to PEM format and write it to a file atomically and with
+  appropriate permissions for a private key."
+  [key :- PrivateKey
+   path :- schema/Str]
+  (ks-file/atomic-write path (partial utils/key->pem! key) private-key-perms))
+
+(schema/defn write-cert
+  "Encode a certificate PEM format and write it to a file atomically and with
```
Comment (mwaggett) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:453]:
```suggestion
  "Encode a certificate to PEM format and write it to a file atomically and with
```

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 2]
```diff
@@ -570,29 +632,37 @@
...
+        (try
+          (let [infra-serial (-> (path-to-cert signeddir infra-node)
+                                 (utils/pem->cert)
+                                 (utils/get-serial))]
+            (.write writer (str infra-serial))
+            (.newLine writer))
+          (catch java.io.FileNotFoundException ex
+            (log/warn
+             (i18n/trs
+              (str
```
Comment (mwaggett) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:649]: I don't think you need to wrap this in `str` - `trs` should let you interpolate with `{0}`.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 3]
```diff
@@ -423,6 +432,51 @@
...
+   path :- schema/Str]
+  (ks-file/atomic-write path (partial utils/objs->pem! crls) public-key-perms))
+
+(schema/defn write-csr
+  "Encode a CSR to PEM format and write it to a file atomically and with
+  appropriate permissions."
+  [csr :- CertificateRequest
+   path :- schema/Str]
+  (ks-file/atomic-write path (partial utils/obj->pem! csr) public-key-perms))
+
```
Comment (Magisus) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:479]: The `*->pem!` functions that these are replacing were all originally defined in the ssl-utils library. Do we think these should actually replace them there at some point? I don't have a strong opinion, but these seem like simply better versions of those.

**Comment by Magisus** [comment]
Really like how neat this turned out!

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj (hunk 3 above)
Comment (pcarlisle) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj:479]: I wasn't sure about this. Maybe at some point?

Comment (Magisus) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj:479]: Yeah, I'm totally fine merging this as is though.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj (hunk 2 above)
Comment (pcarlisle) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj:649]: That's true, but it's fine either way and this isn't code that actually changed in this PR so I'm not going to touch it.

Comment (mwaggett) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj:649]: fair

**Pull request merged**